### PR TITLE
Adds extension method to create service scope that implements IAsyncDisposable.

### LINKED ISSUE #43970: Make CreateScope return ServiceScope that also implements IAsyncDisposable
## Background and Motivation

`IAsyncDisposable` support in DI was added a while ago (see https://github.com/dotnet/extensions/issues/426), but then it was agreed to only add interface implementation to `ServiceProvider`. I didn't find any tracking issue on implementing it also on scopes.

Currently, the following code will throw exception:
```csharp
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;

await using var provider = new ServiceCollection()
        .AddScoped<Foo>()
        .BuildServiceProvider();

using (var scope = provider.CreateScope())
{
    var foo = scope.ServiceProvider.GetRequiredService<Foo>();
}

class Foo : IAsyncDisposable
{
    public ValueTask DisposeAsync() => default;
}
```
```
Unhandled exception. System.InvalidOperationException: 'Foo' type only implements IAsyncDisposable. Use DisposeAsync to dispose the container.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at <Program>$.<<Main>$>d__0.MoveNext() in C:\src\tmp\asyncdisposablescope\Program.cs:line 12
--- End of stack trace from previous location ---
   at <Program>$.<<Main>$>d__0.MoveNext() in C:\src\tmp\asyncdisposablescope\Program.cs:line 12
--- End of stack trace from previous location ---
   at <Program>$.<Main>(String[] args)
```

Workaround today is to cast the returned scope to `IAsyncDisposable`, because the concrete implementation in DI implements that.
```csharp
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;

await using var provider = new ServiceCollection()
        .AddScoped<Foo>()
        .BuildServiceProvider();

{
    var scope = provider.CreateScope();

    var foo = scope.ServiceProvider.GetRequiredService<Foo>();

    await ((IAsyncDisposable)scope).DisposeAsync();
}

class Foo : IAsyncDisposable
{
    public ValueTask DisposeAsync() => default;
}
```

## Proposed API

`IServiceScope` implements `IAsyncDisposable`
```diff
 namespace Microsoft.Extensions.DependencyInjection
 {
-    public interface IServiceScope : IDisposable
+    public interface IServiceScope : IDisposable, IAsyncDisposable
 }
```

## Usage Examples

Changing this would allow us to use `DisposeAsync` on the returned scope.
```csharp
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;

await using var provider = new ServiceCollection()
        .AddScoped<Foo>()
        .BuildServiceProvider();

await using (var scope = provider.CreateScope())
{
    var foo = scope.ServiceProvider.GetRequiredService<Foo>();
}

class Foo : IAsyncDisposable
{
    public ValueTask DisposeAsync() => default;
}
```

## Alternative Designs

Create a new interface `IAsyncDisposableServiceScope` (needs better naming), which implements `IServiceScope` and `IAsyncDisposable` and let `IServiceScopeFactory.CreateScope()` return that.

```diff
 namespace Microsoft.Extensions.DependencyInjection
 {
+    public interface IAsyncDisposableServiceScope : IServiceScope, IAsyncDisposable { }

     public interface IServiceScopeFactory
     {
-        IServiceScope CreateScope();
+        IAsyncDisposableServiceScope CreateScope();
     }

    public static class ServiceProviderServiceExtensions
    {
-        public static IServiceScope CreateScope(this IServiceProvider provider)
+        public static IAsyncDisposableServiceScope CreateScope(this IServiceProvider provider)
    }
}
```

## Risks

Will probably break some 3rd party implementations of `Microsoft.Extensions.DependencyInjection.Abstractions` and consumers.

### PR DESCRIPTION
**Pull request by bjorkstromm**
- Introduces a new type AsyncServiceScope that implements IServiceScope and IAsyncDisposable. The type just wraps an existing IServiceScope instance, which it tries to cast it to an IAsyncDisposable when DisposeAsync is called.
- Adds netstandard2.1 target to avoid bringing in System.Threading.Tasks.Extensions and Microsoft.Bcl.AsyncInterfaces if not needed.
- Fixes #43970

Can't remember (nor find info) whether we agreed on putting this in `Microsoft.Extensions.DependencyInjection` or `Microsoft.Extensions.DependencyInjection.Abstractions`. I added it to `Microsoft.Extensions.DependencyInjection.Abstractions`, which however only targeted `net461` and `netstandard2.0`. This means that we have to pull in `System.Threading.Tasks.Extensions` (for `ValueTask`) and `Microsoft.Bcl.AsyncInterfaces` (for `IAsyncDisposable`). Because of this, I think it would be wise to add a `netstandard2.1` target so that these dependencies can be ignored on `net5` and newer,

// cc @davidfowl

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        /// <inheritdoc />
+        public ValueTask DisposeAsync()
+        {
+            if (_serviceScope is IAsyncDisposable ad)
+            {
+                return ad.DisposeAsync();
+            }
+            _serviceScope.Dispose();
+            return default;
```
Comment (bjorkstromm) [inline src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs:48]: `ValueTask.CompletedTask` is only available in `net5.0` and later.
https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask.completedtask?view=net-5.0

**Comment by davidfowl** [comment]
This looks good but needs tests

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        /// <param name="serviceScope">The <see cref="IServiceScope"/> instance to wrap.</param>
+        /// </summary>
+        public AsyncServiceScope(IServiceScope serviceScope)
+        {
+            if (serviceScope is null)
+            {
+                throw new ArgumentNullException(nameof(serviceScope));
+            }
+
+            _serviceScope = serviceScope;
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs:28]:
```suggestion
            _serviceScope = serviceScope ?? throw new ArgumentNullException(nameof(serviceScope));
```

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.csproj [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="Microsoft.Extensions.DependencyInjection.Abstractions.cs" />
     <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\DynamicallyAccessedMembersAttribute.cs" />
     <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\DynamicallyAccessedMemberTypes.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' or
+                        $([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(SystemThreadingTasksExtensionsVersion)" />
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Bcl.AsyncInterfaces\src\Microsoft.Bcl.AsyncInterfaces.csproj" />
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.csproj:14]: What's this for?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.csproj:14]: Ah `IAsyncDisposable`

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// A <see cref="IServiceScope" /> implementation that implements <see cref="IAsyncDisposable" />.
+    /// </summary>
+    public struct AsyncServiceScope : IServiceScope, IAsyncDisposable
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs:12]:
```suggestion
    public readonly struct AsyncServiceScope : IServiceScope, IAsyncDisposable
```

**Comment by bjorkstromm** [comment]
> 
> 
> This looks good but needs tests



> 
> 
> This looks good but needs tests

Tests added to `Microsoft.Extensions.DependencyInjection.Tests`

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs [hunk 5]
```diff
@@ -19,6 +19,15 @@ public partial class ActivatorUtilitiesConstructorAttribute : System.Attribute
     {
         public ActivatorUtilitiesConstructorAttribute() { }
     }
+    public struct AsyncServiceScope : Microsoft.Extensions.DependencyInjection.IServiceScope, System.IAsyncDisposable
+    {
+        public AsyncServiceScope(Microsoft.Extensions.DependencyInjection.IServiceScope serviceScope)
+        {
```
[the remark this answers is not in the document]
Comment (bjorkstromm) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs:25]: Yes. These was manually added. When running `dotnet msbuild /t:GenerateReferenceAssemblySource` in `src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src` I will get a lot of diffs. Mostly around tooling stripping `System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute`. It seems that the ref source have been manually altered previously. Is this still ok? Example diff can be found here:
https://gist.github.com/bjorkstromm/f128301314db37dbb1274978c6dda193

[the remark this answers is not in the document]
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs:25]: Thanks for the explanation. We don't want to remove those attributes. 

Based on the diff you get from the tool, could you please ignore the attribute diff, and take in only the relevant changes from it as-is?

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs:25]: Is the tool broken?

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs:25]: Yes. But we are re-writing the tool to be Roslyn-based, so we haven't fixed the current version.

cc @safern

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs:43]: Can you stick this in a comment in the code why it is done this way?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 6]
```diff
@@ -895,6 +895,67 @@ public void ProviderScopeDisposeThrowsWhenOnlyDisposeAsyncImplemented()
...
+        public async Task ProviderAsyncScopeDisposeAsyncCallsDisposeAsyncOnServices()
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddTransient<AsyncDisposable>();
+
+            var serviceProvider = CreateServiceProvider(serviceCollection);
+            var scope = serviceProvider.CreateAsyncScope();
+            var disposable = scope.ServiceProvider.GetService<AsyncDisposable>();
+
+            await (scope as IAsyncDisposable).DisposeAsync();
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:908]: Why do you have to cast here and in the next tests?

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs (hunk 1 above)
Comment (bjorkstromm) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs:45]: This is now added.

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjorkstromm) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs]: @maryamariyan I've now merged in changes generated by the tool. There were some diffs in the generated ref source for `AsyncServiceScope`.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs (hunk 6 above)
Comment (bjorkstromm) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs]: Ah, just a case of copy & paste when copying some tests for `CreateScope()`

File: src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (safern) [reply src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/ref/Microsoft.Extensions.DependencyInjection.Abstractions.cs]: > Is the tool broken?

Not broken per se, the tool filters out internal attributes and in this assembly those attributes are compiled as internal. We should however figure out a way to do this better as mostly all internal attributes like compiler generated attributes (nullable, readonly, etc) we don't want to preserve in ref, however we do want to preserve some others like these ones.

**Review by davidfowl — APPROVED** [review]
LGTM

**Review by eerhardt — APPROVED** [review]
Looks good. Thanks for the contribution!

**Comment by davidfowl** [comment]
Excellent work @bjorkstromm Im looking forward to more changes like this

**Pull request merged**